src/cargo/ops/cargo_install.rs

```diff
@@ -85,15 +566,38 @@ pub fn install(
                 no_track,
                 !did_update,
             ) {
-                Ok(still_needs_update) => {
+                Ok(Some(installable_pkg)) => {
+                    did_update = true;
+                    Some((krate, installable_pkg))
+                }
+                Ok(None) => {
+                    // Already installed
                     succeeded.push(krate);
-                    did_update |= !still_needs_update;
+                    None
                 }
                 Err(e) => {
                     crate::display_error(&e, &mut config.shell());
                     failed.push(krate);
                     // We assume an update was performed if we got an error.
                     did_update = true;
+                    None
+                }
+            }
+        });
+
+        let install_results: Vec<_> = pkgs_to_install
+            .into_iter()
+            .map(|(krate, installable_pkg)| (krate, installable_pkg.install_one()))
+            .collect();
```

Comment:
Here's the core change. We're first creating all the `InstallablePackage`, and later calling `install_one` on each of them.